linkerd/app/outbound/src/opaq/logical/route/filters.rs

```diff
@@ -1,30 +1,98 @@
+    pin::Pin,
+    sync::Arc,
+    task::{Context, Poll},
+};
 
-pub(crate) fn apply<T>(t: T) -> Result<T, Error>
+#[derive(Clone, Debug)]
+pub struct NewApplyFilters<N> {
+    inner: N,
+}
+
+#[derive(Clone, Debug)]
+pub struct ApplyFilters<S> {
+    inner: S,
+    filters: Arc<[opaq::Filter]>,
+}
+
+// === impl NewApplyFilters ===
+
+impl<N> NewApplyFilters<N> {
+    pub fn layer() -> impl svc::layer::Layer<N, Service = Self> + Clone {
+        svc::layer::mk(move |inner| Self { inner })
+    }
+}
+
+impl<T, N, S> svc::NewService<T> for NewApplyFilters<N>
 where
-    T: Clone,
+    N: svc::NewService<T, Service = S>,
     T: svc::Param<Arc<[opaq::Filter]>>,
 {
-    let filters: &[opaq::Filter] = &t.param();
-    if let Some(filter) = filters.iter().next() {
-        match filter {
-            opaq::Filter::Forbidden => {
-                return Err(errors::TCPForbiddenRoute.into());
-            }
+    type Service = ApplyFilters<S>;
 
-            opaq::Filter::Invalid(message) => {
-                return Err(errors::TCPInvalidBackend(message.clone()).into());
-            }
+    fn new_service(&self, target: T) -> Self::Service {
+        let filters: Arc<[opaq::Filter]> = target.param();
+        let svc = self.inner.new_service(target);
+        ApplyFilters {
+            inner: svc,
+            filters,
+        }
+    }
+}
 
-            opaq::Filter::InternalError(message) => {
-                return Err(errors::TCPInvalidPolicy(message).into());
+// === impl ApplyFilters ===
+
+impl<S> ApplyFilters<S> {
+    fn apply_filters(&self) -> Result<(), Error> {
+        if let Some(filter) = self.filters.iter().next() {
+            match filter {
+                opaq::Filter::Forbidden => {
+                    return Err(errors::TCPForbiddenRoute.into());
+                }
+
+                opaq::Filter::Invalid(message) => {
+                    return Err(errors::TCPInvalidBackend(message.clone()).into());
+                }
+
+                opaq::Filter::InternalError(message) => {
+                    return Err(errors::TCPInvalidPolicy(message).into());
+                }
             }
         }
+
+        Ok(())
+    }
+}
+
+impl<I, S> svc::Service<I> for ApplyFilters<S>
+where
+    I: io::AsyncRead + io::AsyncWrite + Send + 'static,
+    S: svc::Service<I> + Send + Clone + 'static,
+    S::Error: Into<Error>,
+    S::Future: Send,
+{
+    type Response = S::Response;
+    type Error = Error;
+    type Future = Pin<Box<dyn Future<Output = Result<S::Response, Error>> + Send + 'static>>;
+
+    #[inline]
+    fn poll_ready(&mut self, cx: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
+        self.inner.poll_ready(cx).map_err(Into::into)
     }
 
-    Ok(t)
+    fn call(&mut self, io: I) -> Self::Future {
+        let call = self.inner.call(io);
+        let apply = self.apply_filters();
+
+        Box::pin(async move {
+            apply?;
+            call.await.map_err(Into::into)
+        })
+    }
```

Comment:
Note that this can potentially dispatch the connection even when the filters error--if the call future is spawned on another task, it may progress despite the fact that we don't poll it in the returned future.

Instead, we can do something like:

```rust
    type Future = future::Either<
        future::ErrInto<S::Future, Error>,
        future::Ready<Result<S::Response, Error>>,
    >; 

    fn call(&mut self, io: I) -> Self::Future {
        if let Err(e) = self.apply_filters() {
            return future::Either::Right(future::err(e));
        }
        future::Either::Left(self.inner.call(io).err_into())
    }
```

This also saves us an allocation (Box).